File: Server/AnpanMMOServer/Math/MathUtil.cpp
#include "stdafx.h"
#include "MathUtil.h"
#include "Rotation.h"

// int‚Ì•âŠÔ.
template<>
int MathUtil::Lerp(int Start, int End, float Rate)
{
	if (Rate <= 0.0f)
	{
		Rate = 0.0f;
	}
	if (Rate >= 1.0f)
	{
		Rate = 1.0f;
	}
	return (int)((1.0f - Rate) * Start + Rate * End);
}

// float‚Ì•âŠÔ.
template<>
float MathUtil::Lerp(float Start, float End, float Rate)
{
	if (Rate <= 0.0f)
	{
		Rate = 0.0f;
	}
	if (Rate >= 1.0f)
	{
		Rate = 1.0f;
	}
	return (1.0f - Rate) * Start + Rate * End;
}

// Vector3D‚Ì•âŠÔ.
template<>
Vector3D MathUtil::Lerp(Vector3D Start, Vector3D End, float Rate)
{
	Vector3D Result = Start;
	Result.X += (End.X - Start.X) * Rate;
	Result.Y += (End.Y - Start.Y) * Rate;
	Result.Z += (End.Z - Start.Z) * Rate;
	return Result;
}

// ‰ñ“]‚Ì•âŠÔ.
template<>
Rotation MathUtil::Lerp(Rotation Start, Rotation End, float Rate)
{
	float Value = Lerp<float>(Start.Get(), End.Get(), Rate);
	return Rotation(Value);
}

File: Tool/HeightMapGenerator/MathUtil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HeightMapGenerator
{

	/// <summary>
	/// 計算ユーティリティ
	/// </summary>
	public class MathUtil
	{

		/// <summary>
		/// 補間.
		/// </summary>
		/// <param name="Start">開始</param>
		/// <param name="End">終了</param>
		/// <param name="Rate">レート</param>
		/// <returns>補間結果</returns>
		public static float Lerp(float Start, float End, float Rate)
		{
			if(Rate > 1.0f)
			{
				Rate = 1.0f;
			}
			if(Rate < 0.0f)
			{
				Rate = 0.0f;
			}
			return ((1.0f - Rate) * Start + Rate * End);
		}

	}
}
